### PR TITLE
avoid performing garbage collect on each topology update

### PR DESCRIPTION
**Pull request by murali-reddy**
Fixes: #115

### CODE REVIEW

**Comment by bboreham** [comment]
I think I would be happier if we did a garabgeCollect _somewhere_.
Maybe do it on the gossipTimer in localPeer ?

**Comment by murali-reddy** [comment]
Updated PR to perform garbage collect corresponding to topology updates. Also timer is used trigger garbage collect at scheduled interval which will help coalesce topology updates.

**Comment by murali-reddy** [comment]
addressed review comments. PTAL

File: gossip_test.go [hunk 1]
```diff
@@ -66,6 +67,8 @@ func sendPendingGossip(routers ...*Router) {
 			sentSomething = router.sendPendingGossip() || sentSomething
 		}
 	}
+	// Allow topology to propagate
+	time.Sleep(1 * time.Second)
```
Comment (bboreham) [inline gossip_test.go]: I don't like this - `sendPendingGossip()` is carefully written to wait long enough already.

You could add a new function in `peers_test.go` like:
```go
func forcePendingGC(routers ...*Router) {
	for _, router := range routers {
		router.Peers.Lock()
		if router.Peers.pendingGC {
			var pending peersPendingNotifications
			router.Peers.garbageCollect(&pending)
			router.Peers.unlockAndNotify(&pending)
		} else {
			router.Peers.Unlock()
		}
	}
}
```

then call it as necessary?

**Comment by bboreham** [comment]
I note we can't run `go test -race` on this package any more, because `TestDeferredShortIDReassignment()` needs to create 4096 peers, and each peer now creates two goroutines, and the Go / LLVM ThreadSanitizer has a limit of 8128 goroutines.

**Pull request merged**